File: src/execution/sandbox.ts
import { SSHExecutor, ExecutionResult } from './ssh_client';
import { PathValidator } from '../security/path_validator';

export class SandboxRuntime {
  constructor(
    private ssh: SSHExecutor,
    public pathValidator: PathValidator,
    private isWindows: boolean
  ) {}

  public async execute(command: string, targetDir: string = '.'): Promise<ExecutionResult> {
    const resolvedPath = this.pathValidator.resolveSafe(targetDir);
    
    let sandboxedCmd = '';
    if (this.isWindows) {
      sandboxedCmd = `cd /d "${resolvedPath}" && ${command}`;
    } else {
      sandboxedCmd = `cd "${resolvedPath}" && ${command}`;
    }

    // Wrap output with a boundary
    const wrappedCmd = `echo ========== UNTRUSTED TOOL OUTPUT ========== && ${sandboxedCmd}`;
    
    return this.ssh.execute(wrappedCmd, resolvedPath);
  }

  public async writeFile(filePath: string, content: string): Promise<ExecutionResult> {
    const resolvedPath = this.pathValidator.resolveSafe(filePath);
    const b64 = Buffer.from(content).toString('base64');
    
    let cmd = '';
    if (this.isWindows) {
      // In Windows, use powershell to decode base64
      cmd = `powershell -Command "[IO.File]::WriteAllBytes('${resolvedPath}', [Convert]::FromBase64String('${b64}'))"`;
    } else {
      cmd = `echo "${b64}" | base64 -d > "${resolvedPath}"`;
    }
    
    return this.ssh.execute(cmd, '.');
  }
}

File: legacy_python/src/agent/infrastructure/execution/sandbox.py
from __future__ import annotations

import logging
from pathlib import Path

from agent.domain.schemas import ExecutionResult
from agent.infrastructure.execution.ssh_client import SSHExecutor
from agent.security.path_validator import PathValidator

logger = logging.getLogger(__name__)


class SandboxRuntime:
    """Sandboxed command execution with path validation and confinement."""

    def __init__(
        self,
        ssh: SSHExecutor,
        sandbox_root: Path,
        path_validator: PathValidator,
    ) -> None:
        self._ssh = ssh
        self._sandbox_root = sandbox_root
        self._path_validator = path_validator

    async def execute_sandboxed(
        self,
        command: str,
        working_dir: str = ".",
        timeout: int = 30,
    ) -> ExecutionResult:
        """Execute a command within the validated sandbox boundary.

        Args:
            command: Shell command to execute.
            working_dir: Directory relative to sandbox root.
            timeout: Maximum execution time in seconds.

        Returns:
            ExecutionResult with exit code, output, and duration.
        """
        if not self._path_validator.validate(working_dir):
            logger.warning("Path validation failed for: %s", working_dir)
            return ExecutionResult(
                exit_code=126,
                stdout="",
                stderr=f"Path validation failed: {working_dir}",
                duration_ms=0.0,
                truncated=False,
            )

        resolved = self._path_validator.resolve_safe(working_dir)
        full_path = resolved

        # For Windows CMD, cd needs the /d flag to switch drives if we provide an absolute path
        if self._path_validator.is_windows:
            sandboxed_cmd = f'cd /d "{full_path}" && {command}'
        else:
            sandboxed_cmd = f'cd "{full_path}" && {command}'

        logger.info(
            "Sandboxed execution in %s: %.80s",
            full_path,
            command,
        )

        return await self._ssh.execute(
            command=sandboxed_cmd,
            working_dir=full_path,
            timeout=timeout,
        )
